Significantly improve tight layout performance for cartopy axes

lukelbd opened the pull request:
This PR is adapted from SciTools/cartopy#1956 following our discussion there.

## Summary

This PR significantly improves the speed of `Axes.get_tightbbox()` for non-rectilinear axes (in particular, matplotlib's PolarAxes and cartopy's GeoAxes) by preventing unnecessary and expensive `get_window_extent()` computations. In the presence of complex, high-resolution artists, it can result in a **2x improvement** to the draw time when "tight layout" is enabled (since artists are effectively only drawn once instead of twice).

There may be a better way to implement this -- looking forward to everyone's thoughts.

## Details

To prevent unnecessary and expensive `get_window_extent()` computations, `Axes.get_tightbbox()` skips artists with `clip_on` set to True and whose `clip_box.extents` are equivalent to `ax.bbox.extents`. However, while all axes artists are clipped by `TransformedPatchPath(ax.patch)` by default, only artists drawn inside rectilinear projections are clipped by `ax.bbox` (see the below example).

This PR replaces `Artist._get_clipping_extent_bbox()` with `Artist._is_axes_clipped()`. Now,`Axes.get_tightbbox()` includes `ax.patch.get_window_extent()` in its computation, and skips all artists clipped by **either** `ax.bbox` or `ax.patch` (i.e., artists for which `Artist._is_axes_clipped()` returns True).

This PR also removes the computation of the intersection of `clip_box` and `clip_path`, under the assumption that the independent test of `clip_path` covers those instances, but perhaps that should be added back.

## Example

Here is an example with a cartopy `GeoAxes`:

```python
import numpy as np
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
fig, ax = plt.subplots(subplot_kw={'projection': ccrs.Robinson()})
N = 5000  # large dataset
lon = np.linspace(-180, 180, N)
lat = np.linspace(-90, 90, N)
data = np.random.rand(N, N)
m = ax.pcolormesh(lon, lat, data, transform=ccrs.PlateCarree())
print(m.get_clip_box() is None)  # returns True
%timeit fig.tight_layout()
```

Performance before:

```
638 ms ± 67.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Performance after:

```
2.37 ms ± 60.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

And of course the performance difference is larger the larger the dataset.

## Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] (n/a?) New features are documented, with examples if plot related.
- [x] (n/a?) New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] (n/a?) API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

lukelbd commented:
Added a simple test that `artist._is_axes_clipped()` returns `True` under default conditions / various projections. Note that all non-rectilinear axes (e.g. `PolarAxes`, `MollweideAxes`, cartopy `GeoAxes`) have `artist.get_clip_box()` set to `None`.

lukelbd commented:
I can also see the origin of this bug: The `Artist._get_clipping_extent_bbox()` lines that I removed (previously used inside `Axes.get_tightbbox()` to skip artists) mirror the below lines in `Artist.get_tightbbox()`.

https://github.com/matplotlib/matplotlib/blob/1a74793837286b588789441c6285c174f643a5ea/lib/matplotlib/artist.py#L340-L364

While `Artist.get_tightbbox()` starts the `if` block with a valid `bbox`, `Artist._get_clipping_extent_bbox()` starts with `bbox = None`... so when it gets to Line 361, the `clip_path` is ignored, because there is no `bbox` to intersect with.

jklymak commented on lib/matplotlib/artist.py:
```diff
@@ -320,22 +320,34 @@ def get_window_extent(self, renderer):
+                return False
+        if clip_path is not None:
+            if not isinstance(clip_path, TransformedPatchPath):
+                return False
```
I think it would be good if test coverage were 100% on this PR....

lukelbd replied:
I now include a test that `_fully_clipped_by_axes` returns `False` for non-default clipping situations.

tacaswell commented on lib/matplotlib/artist.py:
```diff
@@ -844,6 +827,30 @@ def get_in_layout(self):
         """
         return self._in_layout
 
+    def _get_skip_layout(self):
```
Maybe call this `_fully_clipped_to_axes` or something similarly more descriptive as to _why_ we want to skip it in the layout.  There are many reasons a user could imagine you want to skip the including an Artists in the layout, we want to avoid the case where each user imagines that this method implements _their_ favorite justification ;)

The name should also hint that this only matters for Axes and not for figure level layouts.

lukelbd replied:
Done!

lukelbd commented:
I've improved this PR in a force push:

* Renamed `_is_axes_clipped()` to `_fully_clipped_to_axes` (thanks @tacaswell).
* Added test lines that assert `_get_skip_layout()` is False for non-default clipping situations (re: @jklymak).
* Removed the redundant addition of `ax.patch` to the bboxes (it's included in `get_default_bbox_extra_artists()`).
* Condensed multiple return statements into a single boolean return statement (possibly cleaner?)
* Moved the line that skips artists from `Axes.get_tightbbox()` to `Axes.get_default_bbox_extra_artists()`. Note this means the public `get_default_bbox_extra_artists()` will return different values compared to previous versions.

I also made the following additional changes:

* Removed an unnecessary `bbox is not None` check in `Artist.get_tightbbox()` (this is always True).
* Removed axis instances from `get_default_bbox_extra_artists()` (this was redundant -- see `Axes.get_tightbbox()`).
* Replaced `Axes.get_tightbbox()` reference to `[xy]axis` with `ax._get_axis_list()` (consistent with removed lines).
* Always include artists that don't internally implement clipping (should fix obscure tight layout "bugs").

The latter points also resolve a second inefficiency: Currently, `Axes.get_tightbbox()` *both* 1) computes the `xaxis` and `yaxis` tight bbox using `for_layout_only=True` on these lines:

https://github.com/matplotlib/matplotlib/blob/710fce3df95e22701bd68bf6af2c8adbc9d67a79/lib/matplotlib/axes/_base.py#L4616-L4634

and 2) includes `xaxis` and `yaxis` as "extra artists" in `get_default_bbox_extra_artists()`. This meant that their bounding boxes would be effectively calculated twice if they were not "skipped" by the `_get_clipping_extent_bbox` check. For some reason, this check *was* successfully skipping `xaxis` and `yaxis` in rectilinear axes, because they have a `clip_on` set to `True` and `clip_box` set to `ax.bbox` (despite the fact that they are *not* clipped when drawn..... weird). However, this failed to skip non-rectilinear axis instances, because they only have a `clip_path` set to `ax.patch`. After this PR, axis instances are excluded from `Axes.get_default_bbox_extra_artists()`, and their extents are always calculated once.

I've also verified that, as before, `ax.patch` and `ax.spines` are still included as default "extra artists". Here's a test I used to determine the default clipping settings for various axes components:

```python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
fig = plt.figure(figsize=(4, 2))
ax1 = fig.add_subplot(131)
ax2 = fig.add_subplot(132, projection='polar')
ax3 = fig.add_subplot(133, projection='hammer')
for ax in (ax1, ax2, ax3):
    print(ax)
    for a in (ax.patch, tuple(ax.spines.values())[0]):
        clip_box = a.get_clip_box()
        b1 = clip_box is not None and np.all(clip_box.extents == ax.bbox.extents)
        clip_path = a.get_clip_path()
        b2 = clip_path is not None and clip_path._patch is ax.patch
        print(a, '\nclip on?', a.get_clip_on(), '\naxes clip box?', b1, '\naxes clip path?', b2, '\nFULLY CLIPPED?', a._fully_clipped_to_axes())
    print()
```

Results:

```python
AxesSubplot(0.125,0.11;0.227941x0.77)
Rectangle(xy=(0, 0), width=1, height=1, angle=0)
clip on? True
axes clip box? False
axes clip path? False
FULLY CLIPPED? False
Spine
clip on? True
axes clip box? False
axes clip path? False
FULLY CLIPPED? False

PolarAxesSubplot(0.398529,0.11;0.227941x0.77)
Wedge(center=(0.5, 0.5), r=0.5, theta1=0, theta2=360, width=None)
clip on? True
axes clip box? False
axes clip path? False
FULLY CLIPPED? False
Spine
clip on? True
axes clip box? False
axes clip path? False
FULLY CLIPPED? False

HammerAxesSubplot(0.672059,0.11;0.227941x0.77)
Circle(xy=(0.5, 0.5), radius=0.5)
clip on? True
axes clip box? False
axes clip path? False
FULLY CLIPPED? False
Spine
clip on? True
axes clip box? False
axes clip path? False
FULLY CLIPPED? False
```

jklymak commented:
This all looks good, but are we sure it has similar performance to before?  https://github.com/matplotlib/mpl-bench has benchmarking mechanisms, in particular you could use the constrained_layout and tight_layout tests?

lukelbd commented:
Here are some very quick benchmarks with 2-column figures of cartopy axes (thanks for pointing me to `mpl-benchmark` -- don't have time to look into that now, but will later). Repeating the tests several times gives qualitatively similar results. If I have time later I can post more rigorous benchmarks.

In principle, this fits as a matplotlib PR, but practically, the speedup *does* seem to be cartopy-specific. I get essentially zero speedup for 1) non-rectilinear matplotlib axes and 2) by omitting the `transform=ccrs.PlateCarree()` line in the cartopy example. It seems that for high-res datasets, the limitation in cartopy is translating coordinates (triggered in both `get_window_extent` and `draw`). In those cases, this PR can deliver a roughly **2x** speedup.

## Cartopy JPG test

Test code:

```python
import time
import numpy as np
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import proplot as pplt
fig, axs = plt.subplots(ncols=2, figsize=(8, 4), subplot_kw={'projection': ccrs.Stereographic()})
t1 = time.time()
lon = np.linspace(0, 360, 3000)
lat = np.linspace(-60, 60, 500)
state = np.random.RandomState(51423)
data = state.rand(len(lat) - 1, len(lon) - 1)
for ax in axs:
    ax.coastlines()
    ax.pcolormesh(lon, lat, data, transform=ccrs.PlateCarree())
t2 = time.time()
print('Plot time:', t2 - t1)
fig.savefig('test.jpg', dpi=300, bbox_inches=None)
# fig.savefig('test.jpg', dpi=300, bbox_inches='tight')
t3 = time.time()
print('Save time:', t3 - t2)
```

Results with `bbox_inches=None`:

```python
Plot time: 5.106376886367798
Save time: 4.24917197227478
```

Results with `bbox_inches='tight'`, *after* this PR (save time ~20% slower):

```python
Plot time: 5.4417970180511475
Save time: 5.905539035797119
```

Results with `bbox_inches='tight'`, *before* this PR (save time ~120% slower):

```python
Plot time: 5.133382081985474
Save time: 10.721824884414673
```

## Cartopy PDF tests

Test code:

```python
import time
import numpy as np
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import proplot as pplt
fig, axs = plt.subplots(ncols=2, figsize=(8, 4), subplot_kw={'projection': ccrs.Stereographic()})
t1 = time.time()
lon = np.linspace(0, 360, 100)
lat = np.linspace(-60, 60, 100)
state = np.random.RandomState(51423)
data = state.rand(len(lat) - 1, len(lon) - 1)
for ax in axs:
    ax.coastlines()
    ax.pcolormesh(lon, lat, data, transform=ccrs.PlateCarree())
t2 = time.time()
print('Plot time:', t2 - t1)
fig.savefig('test.pdf')
t3 = time.time()
print('Save time:', t3 - t2)
```

Results with `bbox_inches=None`:

```python
Plot time: 0.28498291969299316
Save time: 2.3959097862243652
```

Results with `bbox_inches='tight'`, *after* this PR (save time ~10% slower):

```python
Plot time: 0.24462389945983887
Save time: 2.699641227722168
```

Results with `bbox_inches='tight'`, *before* this PR (save time ~60% slower):

```python
Plot time: 0.2555239200592041
Save time: 3.705993890762329
```

## Matplotlib JPG test

Test code:

```python
import time
import numpy as np
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import proplot as pplt
fig, axs = plt.subplots(ncols=2, figsize=(8, 4), subplot_kw={'projection': 'hammer'})
t1 = time.time()
lon = np.linspace(0, 360, 500)
lat = np.linspace(-60, 60, 500)
state = np.random.RandomState(51423)
data = state.rand(len(lat) - 1, len(lon) - 1)
for ax in axs:
    ax.pcolormesh(lon, lat, data)
t2 = time.time()
print('Plot time:', t2 - t1)
fig.savefig('test.jpg', dpi=300, bbox_inches='tight')
t3 = time.time()
print('Save time:', t3 - t2)
```

Results with `bbox_inches=None`:

```python
0.04705810546875
Save time: 12.776510953903198
```

Results with `bbox_inches='tight'`, *after* this PR (save time about the same):

```python
Plot time: 0.05183219909667969
Save time: 11.54852819442749
```

Results with `bbox_inches='tight'`, *before* this PR (save time about the same):

```python
Plot time: 0.053359031677246094
Save time: 11.422415018081665
```

## Summary

As noted in the original cartopy thread (SciTools/cartopy#1956) this applies for all hi-res artists, not just quad meshes. The problem also seems to be worse for bitmap formats (jpg or png are usually a **2x** speedup) rather than vector formats (pdf is a **10%** speedup).

My current mental model for this problem is "for high-res data, `.get_window_extent()` is approximately as expensive as `.draw()`, because the limiting step is translating the dense vector paths into bitmap." Though not sure how accurate this is.

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -4541,21 +4541,26 @@ def get_default_bbox_extra_artists(self):
-                if (artist.get_visible() and artist.get_in_layout())]
+        # always include types that do not internally implement clipping
+        # to axes. may have clip_on set to True and clip_box equivalent
+        # to ax.bbox but then ignore these properties during draws.
```
That is .... a fun fact.
